Move visualization properties from visualizer to Nodes

dcambie opened the pull request:
For many PvTrace simulation it would be nice to be able to add node visualization properties,
such as color transparency etc.
However, currently, the only visualization options (wireframe, transparency, opacity and reflectivity)
are attributes of MeshcatRenderer and apply in the same way to all the scene objects.

This can be modified at runtime *by setting the corresponding Meshcat material attributes)
but there is currently no way to save the visualization options in the scene definition (e.g. in
the YML file)

This commit allows for more granular control on the visual appearances of the scene, when rendered.
At the same time, the old global switches at the MeshcatRenderer are kept and are given priority as,
e.g. it may be needed/necessary/useful to show all scene nodes as wireframe for debugging purposes.

dcambie committed:
Move visualization properties from visualizer to Nodes

For many PvTrace simulation it would be nice to be able to add node visualization properties,
such as color transparency etc.
However, currently, the only visualization options (wireframe, transparency, opacity and reflectivity)
are attributes of MeshcatRenderer and apply in the same way to all the scene objects.

This can be modified at runtime *by setting the corresponding Meshcat material attributes)
but there is currently no way to save the visualization options in the scene definition (e.g. in
the YML file)

This commit allows for more granular control on the visual appearances of the scene, when rendered.
At the same time, the old global switches at the MeshcatRenderer are kept and are given priority as,
e.g. it may be needed/necessary/useful to show all scene nodes as wireframe for debugging purposes.

dcambie committed:
Rollback accidental changes

danieljfarrell commented:
Hi Dario,

Oh great to see you have been testing out the CLI branch. Do you find yourself using this now rather than Python scripts? It's such a big change that I have not merged it into the master yet. But maybe there could be a beta release at some point? Which reminds we, I never got around to publishing the last release, I should do that soon!

---

I really like this idea! 

The history of the visualiser is that I needed something for debugging the ray tracing algorithm and testing all of the node coordinate system transformations. If I remember correctly, I think you actually found a bug in that code, using the visualiser. So it is certainly a bit barebones and I've toyed with the idea of making it an optional dependency, but it does seem popular.

I wonder if visualisation properties/styles could be implemented in away that preserves the strict separation between _model_ and _view_ objects?

I see two places in the pull request where model and view objects have blurry responsibilities:

1. The _model_ class **Material** inheriting from the _view_ class **g.MeshBasicMaterial** could be a problem if we want to add more fully featured visualisers in the future such as VTK or OpenCSG?
2. Similarly, the YML file format adds view attributes directly to the model.

Let me know what you think.

dcambie committed:
Stub of appearance properties in Node

dcambie commented:
Hi Dan,

Ok, I understand your point of keeping model and view separate, it makes a lot of sense.

Would about the visualization attribute being kept as a dictionary in `Node`? I believe that makes more sense than Geometry (whose base class is an abstract base class).

Ideally a few basic appearance attribute that can be shared across all different visualization engines (i.e. `color: int`, `transparency: float` and maybe `visible: bool`) and other could be implementation-specific. Also, it would be nice if any arbitrary parameter can be set for a specific visualization engine and passed directly to the corresponding visualization object without redirection/dedicated PvTrace code.

imho All of this could be as simple as an `appearance` dictionary as new attribute in Node (see pseudocode as last commit)

e.g. corresponding YML (in the schema meshcat and other visualizer can be optional with `"additionalProperties": True`
```yml

  ball-lens:
    location: [0, 0, 2]
    sphere:
      radius: 1.0
      material:
        refractive-index: 1.5
    appearance:
      color: 0x00ff00
      visible: True
      meshcat:
        opacity: 0.5
```

What do you think about that?


Yes I've been playing a bit with the CLI branch as we currently have a script that generate a scene, but for a different reactor it needs some changes so I was considering generating the YML files instead and use those for the scene definition. I'll test more of that in the coming days and I'll come back to you with more details.

danieljfarrell commented:
Yeah this is nice solution. Busy right now but will get back to you later.

Commit:
Adds appearance to node in JSON schema

(Still looking for a better name, but at least serves as placeholder for now)

Commit:
Make schema slightly more strict

not allowing additional properties in appearance

Commit:
Properly implement appearance as per #52

Commit:
Rollback docstring update

Commit:
Minor comment change

Commit:
Fix typo that prevented components to be applied to materials

Commit:
Remove debug print

Commit:
Actually use component in example YML

(It would be nice to also use Absorber and Scatterer, WIP)

Commit:
Bugfix: if coefficient is int _abs_dist is not set!

Commit:
Add absorber to YML example

Commit:
Properly support int as absorption coefficients

Because if float are fine, so should be int.
Rationale similar to type hint (PEP484) stating
"when an argument is annotated as having type float, an argument of type int is acceptable"

Commit:
More lenient duck-typing, np.number vs np.float

dcambie committed:
Bugfix loading file spectra as pd.df

When absorption_spectrum is a numpy ndarray it cannot be evaluated as boolean
(Truth value of array needs any bla bla..)

dcambie committed:
Typo :D

dcambie committed:
Properly handle spectra if ndarray

dcambie committed:
Move location to Node definition in hello world scene

dcambie committed:
Provide more diagnostic error on file non existing

dcambie committed:
Remove debug print

dcambie committed:
Revert changes to sphere_with_luminescent_dye.yml

dcambie committed:
Reduce PR changes

danieljfarrell commented:
Hi Dario,

Let me know when you are ready for me to take a look and I’ll do the merge.

dcambie commented:
I'd say it is ready.

There are just a couple of changes that are not strictly related to this PR but should be non controversial, namely:

- `elif absorption_spectrum is not None:`
these are just to fix the parsing of yml where the spectrum is defined as pandas dataframe. Otherwise when spectrum is a numpy array a `ValueError` is raised (`The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`)

- a more diagnostic error on file not found for spectrum
- a minor change to the hello world.

Commit:
Prevent ValueError

The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

Commit:
Prevent FileNotFoundError by copying to installation dir the json files

The json file was not installed upon pip install.
The fact that if pip install -e is used a symbolic link is created has previously hidden this issue.

FileNotFoundError: [Errno 2] No such file or directory: '[omissis...]\\site-packages\\pvtrace\\cli\\pvtrace-schema-scene-spec.json'

Commit:
Merge remote-tracking branch 'dcambie/material_changes' into material_changes

Commit:
Improve parse() type hinting

parse() would happily accept a pathlib.Path() object as filename, yet its type hinting
suggest only str are valid, resulting in type warning e.g. in PyCharm.
Since filename is directly passed to open() any Pathlike object is fine, so the signature
was updated according to PEP519-suggested[1] hint for all acceptable path-representing types.

(changing this as parse() is an outside-facing API part of pvtrace that is likely to be called
by external code where typing info are valuable)

 [1] https://www.python.org/dev/peps/pep-0519/#provide-specific-type-hinting-support

Commit:
Add DB schema to package_data

Thsi prevent "FileNotFoundError: Cannot find database schema"on pvtrace-cli

Commit:
Add typer and asciiplotlib to install require/requirements

Used in CLI

Commit:
Remove debug print

Commit:
Comment out all debug prints in parse

(For scene with many components this becomes quite verbose)

danieljfarrell commented:
I should make an alpha release of the CLI branch soon!

Commit:
BUGFIX: fix material setter in Cylinder and Sphere

This went unnoticed for long since changing material after object creation is rare.

Commit:
BUGFIX Fix typo

Commit:
Add support for reactor type components

Commit:
Expose parse in pvtrace root

This enables from '''pvtrace import parse''' which is likely to be the only function of parse.py needed externally
